language/diem-vm/parallel-executor/src/outcome_array.rs

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    errors::{Error, Result},
+    task::ExecutionStatus,
+};
+use once_cell::sync::OnceCell;
+
+pub(crate) struct OutcomeArray<T, E> {
+    // Hold the execution results for each individual transactions. Each cell should be set exactly
+    // once.
+    results: Vec<OnceCell<ExecutionStatus<T, Error<E>>>>,
+}
+
+impl<T: Default, E: Send> OutcomeArray<T, E> {
+    pub fn new(len: usize) -> OutcomeArray<T, E> {
+        OutcomeArray {
+            results: (0..len).map(|_| OnceCell::new()).collect(),
+        }
+    }
+
+    pub fn set_result(&self, idx: usize, res: ExecutionStatus<T, Error<E>>) {
+        // Only one thread can write at the time, so just set it.
+
+        let entry = &self.results[idx];
+        assert!(entry.set(res).is_ok());
+    }
+
+    pub fn get_all_results(self, valid_length: usize) -> Result<Vec<T>, E> {
+        let len = self.results.len();
+        let mut results = self
+            .results
+            .into_iter()
+            .take(valid_length)
+            .map(|cell| cell.into_inner())
+            .collect::<Option<Vec<_>>>()
+            .ok_or(Error::InvariantViolation)?;
+
+        match results.pop() {
```

Comment:
But I'm not sure how much you care about your invariant violations. This feels like it could be a lot simpler just written as something like
```
    let mut final_results = Vec::with_capacity(len);
    for (idx, status) in results.into_iter().enumerate() {
        let t = match status {
            ExecutionStatus::Success(t) => t,
            ExecutionStatus::SkipRest(t) if idx == len - 1 => t,
            ExecutionStatus::SkipRest(_) => return Err(Error::InvariantViolation),
            ExecutionStatus::Abort(err) => return Err(err),
            ExecutionStatus::Retry(_) => return Err(Error::InvariantViolation),
            _ => return Err(Error::InvariantViolation),
        };
        final_results.push(t)
    }
    assert!(final_results.len() == len);
    Ok(final_results)
```

Reply:
Thanks!! This is a much cleaner logic!